### ISSUE TITLE
Missed unconditional recursion warning in Default trait implementation.

### ISSUE THREAD

**Issue by a2aaron** [rust-lang/rust#83534]
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=97f86e28bd7cd07f59f1507f35a9c7a3

```rust
pub struct Foo {
    pub x: f32,
    pub y: f32,
}

impl Default for Foo {
    fn default() -> Self {
        Foo {
            x: 1.0f32.floor(),
            ..Default::default()
        }
    }
}
```

The current output is:

```
   Compiling playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 0.83s
```

Ideally the output should look like:

```
   Compiling playground v0.0.1 (/playground)
warning: function cannot return without recursing
  --> src/lib.rs:7:5
   |
7  |     fn default() -> Self {
   |     ^^^^^^^^^^^^^^^^^^^^ cannot return without recursing
...
10 |             ..Default::default()
   |               ------------------ recursive call site
   |
   = note: `#[warn(unconditional_recursion)]` on by default
   = help: a `loop` may express intention better if this is on purpose
```

The above code should produce a warning because it is impossible for `Foo::default()` to return without recursing. This likely occurs because the compiler is conservative when it comes to function calls (ex: it's possible that `1.0f32.floor()` could panic and hence never return) but this should still be rejected because any path that actually would return would require unconditional recursion.

Note that this does not occur if `x` doesn't make a function or method call. For example, this code properly warns about unconditional recursion.

```rust
impl Default for Foo {
    fn default() -> Self {
        Foo {
            x: 1.0,
            ..Default::default()
        }
    }
}
```

I ran into this accidentally when implementing default.

**Comment by ShikChen** [comment]
It's a common trap... #78474 was also for this issue.

**Comment by ehuss** [comment]
Closing as a duplicate of #78474.
